Add Bazel build scripts, source code linting, and IDE integration

jmmv opened the pull request:
This set of commits adds a Bazel build for sandboxfs and then uses that to add source code linting (with pre-commit checks) and IDE integration with VSCode.

These commits are "clean" on their own, so they shall be merged without squashing.

rupertks commented on INSTALL.md:
```diff
@@ -0,0 +1,43 @@
+        and remove the `go_version = "host"` parameter from the
+        `go_register_toolchains` statement.
+
+1.  Run `bazel run //admin/install` to download all required dependencies,
```
The install failed for me (on a Mac). Do we expect users to specify a custom prefix, or to have to use sudo, or ?

```
$ bazel run //admin/install
INFO: Found 1 target...
INFO: From SkylarkAction external/go_stdlib_darwin_amd64_cgo/bin/go [for host]:
# runtime/cgo
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-gno-record-gcc-switches' [-Wunused-command-line-argument]
...
INFO: Elapsed time: 43.372s, Critical Path: 25.90s

INFO: Running command line: bazel-bin/admin/install/install
2017/11/20 13:08:24 cmd/sandboxfs/sandboxfs -> /usr/local/bin/sandboxfs (mode -rwxr-xr-x)
ERROR: failed to copy cmd/sandboxfs/sandboxfs to /usr/local/bin/sandboxfs: cannot open /usr/local/bin/sandboxfs for write: open /usr/local/bin/sandboxfs: permission denied
ERROR: Non-zero return code '1' from command: Process exited with status 1.
````

jmmv replied:
Oh right... yes, you need root to copy onto `/usr/local` obviously, but `sudo bazel` is a pretty bad idea. I've updated the instructions accordingly.

Thanks for trying this instead of just reviewing the code! :-)

rupertks replied:
Thanks!

I just tried following the new INSTALL.md, here's the new error I got:

```
$ bazel build //admin/install
INFO: Found 1 target...
Target //admin/install:install up-to-date:
  bazel-bin/admin/install/install
INFO: Elapsed time: 12.385s, Critical Path: 2.08s

$ ./bazel-bin/admin/install/install
2017/11/22 14:34:18 cmd/sandboxfs/sandboxfs -> /usr/local/bin/sandboxfs (mode -rwxr-xr-x)
ERROR: failed to copy cmd/sandboxfs/sandboxfs to /usr/local/bin/sandboxfs: cannot open cmd/sandboxfs/sandboxfs for read: open cmd/sandboxfs/sandboxfs: no such file or directory

$ sudo ./bazel-bin/admin/install/install
Password:
2017/11/22 14:34:24 cmd/sandboxfs/sandboxfs -> /usr/local/bin/sandboxfs (mode -rwxr-xr-x)
ERROR: failed to copy cmd/sandboxfs/sandboxfs to /usr/local/bin/sandboxfs: cannot open cmd/sandboxfs/sandboxfs for read: open cmd/sandboxfs/sandboxfs: no such file or directory
```

jmmv replied:
OK... let's try again.

I've pushed a fix -- as a new commit to avoid the problem you observed before in GitHub and not being able to see what changed. However, this new commit will need fixing up into the right commit pre-merge.

rupertks replied:
Okay, that worked!

```
$ sudo ./bazel-bin/admin/install/install
Password:
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/cmd/sandboxfs/sandboxfs -> /usr/local/bin/sandboxfs (mode -rwxr-xr-x)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/AUTHORS -> /usr/local/share/doc/sandboxfs/AUTHORS (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/CONTRIBUTING.md -> /usr/local/share/doc/sandboxfs/CONTRIBUTING.md (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/CONTRIBUTORS -> /usr/local/share/doc/sandboxfs/CONTRIBUTORS (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/INSTALL.md -> /usr/local/share/doc/sandboxfs/INSTALL.md (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/LICENSE -> /usr/local/share/doc/sandboxfs/LICENSE (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/NEWS.md -> /usr/local/share/doc/sandboxfs/NEWS.md (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/README.md -> /usr/local/share/doc/sandboxfs/README.md (mode -rw-r--r--)
2017/11/25 19:33:24 bazel-bin/admin/install/install.runfiles/sandboxfs/cmd/sandboxfs/sandboxfs.1 -> /usr/local/share/man/man1/sandboxfs.1 (mode -rw-r--r--)
```

rupertks replied:
Works in Linux too (that was OS X above)

rupertks commented:
The install works for me now, but I couldn't actually run sandboxfs afterwards - raised https://github.com/bazelbuild/sandboxfs/issues/8

rupertks commented on admin/install/install.go:
```diff
@@ -0,0 +1,121 @@
+
+// doInstall performs the installation of all given data files into the prefix.
+func doInstall(sourceDir string, prefix string, data []installSpec) error {
+	for _, spec := range data {
```
installData instead of data?

or filesToInstall ?

jmmv replied:
`installData` might be nice, but I'm wary of shadowing global symbols because that can lead to more confusion. `filesToInstall` sounds nice too, but then I'd also want to rename the global to match this new name so the same issue would arise.

I feel that `data` here is not too bad given that the type definition is immediately before use, the type is descriptive enough, and this is a private function (so the text wouldn't end up in godoc).

Don't feel too strongly about this, so if you have a better suggestion to avoid the shadowing issue...

rupertks replied:
Personally I would rename both to filesToInstall and move the global version of filesToInstall into main

jmmv replied:
Done.

rupertks commented on admin/install/install.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// installData specifies the files to be installed by this tool.
+var installData = []installSpec{
```
Does installData need to be global?

jmmv replied:
It doesn't *have* to be, but I considered this to be the program's data and thus felt better represented as a constant. (It cannot be a constant, though, because constants in Go can only be of primitive types.)

rupertks commented on configure:
```diff
@@ -0,0 +1,85 @@
+# to run Bazel.  (E.g. it could be made to read query outputs from a dependency
+# and then have a Bazel genquery rule to supply such data as a dependency.)
+
+set -e
```
set -euo pipefail?

rupertks replied:
(Just bumping this because GitHub says the previous review was dismissed)

Is it not safe to use set -euo pipefail for the configure script?

jmmv replied:
Wut? I'm sure I typed a reply here!

`set -u`: This would be nice... but the ancient bash (which is `/bin/sh`) shipped on macOS has bugs that cause it to complain on references to `"${@}"` when there are no arguments — and references to empty `"${@}"` should be valid. This makes the call to main, and the argument processing, to fail when this is enabled.

`set -o pipefail`: This is a bash-only feature and I don't want to rely on bash here.

jmmv commented:
I've fixed up the last commit as promised into the corresponding commit and also done a couple of changes to the `configure` script (homogenized global variable naming and added `readonly` qualifiers to them).

These changes require reapproval for merge. The only differences should be in the `configure` script.

Thanks for your patience! I'll try to stay to single-commit pull requests in the future so that things are easier to deal with in GitHub.

rupertks commented:
Just added a few more comments, although they don't show up at the bottom of the timeline on this page. GitHub seems... not great for reviews :)
